pymc/model.py

```diff
@@ -1163,14 +1194,31 @@ def set_data(
                             f"{new_length}, so new coord values for the {dname} dimension are required."
                         )
                 if isinstance(length_tensor, TensorConstant):
+                    # The dimension was fixed in length.
+                    # Resizing a data variable in this dimension would
+                    # definitely lead to shape problems.
                     raise ShapeError(
                         f"Resizing dimension '{dname}' is impossible, because "
                         "a 'TensorConstant' stores its length. To be able "
                         "to change the dimension length, pass `mutable=True` when "
                         "registering the dimension via `model.add_coord`, "
                         "or define it via a `pm.MutableData` variable."
                     )
+                elif isinstance(length_tensor, ScalarSharedVariable):
+                    # The dimension is mutable, but was defined without being linked
+                    # to a shared variable. This is allowed, but slightly dangerous.
+                    warnings.warn(
```

Comment:
I think that the key thing that is missing here is that users can also pass `coords` when they `set_data`. Ideally, we should use the supplied `coords` internally to call `set_dim` and then update the values of the `MutableData` instance. This warning should only be raised if the users haven't passed `coords` along to `set_data`.

Reply:
Hmm.. You're welcome to try, but I think this could become more complicated than the current way with the warning.
The `set_data` method doesn't actually care if the `ScalarSharedVariable` corresponds to a dimension in `Model.dim_lengths` or not!

Most important will be to add test cases that showcase the intended call order

Reply:
What would be the expected flow to change data and make predictions? Before we had to do:

```python
with pm.Model(coords={"A": range(10)}) as m:
    x = pm.MutableData("x", x_values, dims="A")
    y = pm.MutableData("y", y_values, dims="A")
    a = pm.Normal("a", 0, 1)
    b = pm.Normal("b", 0, 1)
    c = pm.HalfNormal("c", 1)
    obs = pm.Normal("obs", mu=a + b * x, sigma=c, observed=y)
    idata = pm.sample()

with m:
    pm.set_data({"x": np.linspace(-2, 3, 100), "y": np.full(100, np.nan)})
    ppc = pm.sample_posterior_predictive(idata)
```

I would like us to be able to do this:
```python
with pm.Model(coords={"A": range(10)}) as m:
    x = pm.Data("x", x_values, dims="A")
    y = pm.Data("y", y_values, dims="A")
    a = pm.Normal("a", 0, 1)
    b = pm.Normal("b", 0, 1)
    c = pm.HalfNormal("c", 1)
    obs = pm.Normal("obs", mu=a + b * x, sigma=c, observed=y)
    idata = pm.sample()

with m:
    pm.set_data({"x": np.linspace(-2, 3, 100), "y": np.full(100, np.nan)}, coords={"A": range(10, 110)})
    ppc = pm.sample_posterior_predictive(idata)
```

But due to the choice of default Mutable/Constant coords, we need to:

1. Manually `add_coord` for "A" and say it is mutable
2. Call `set_dim` before calling `set_data`

Or we have to create the coordinate values using `MutableData`, which I don't really know how to do.